src/net.cpp

```diff
@@ -340,71 +340,160 @@ bool CConnman::CheckIncomingNonce(uint64_t nonce)
-            // name catch this early.
-            LOCK(cs_vNodes);
-            CNode* pnode = FindNode((CService)addrConnect);
-            if (pnode)
-            {
-                pnode->MaybeSetAddrName(std::string(pszDest));
-                CloseSocket(hSocket);
-                LogPrintf("Failed to open new connection, already connected\n");
-                return NULL;
+    if (!strDest.empty()) {
+        // Resolve strDest, if set. Only actually perform a dns lookup if they're enabled
+        // by the current settings. Note that if strDest is already an ip address, this will
+        // succeed regardless of the lookup setting.
+        std::vector<CService> addrResolved;
+        if (Lookup(strDest.c_str(), addrResolved, port, fNameLookup && !HaveNameProxy(), 256)) {
+            if (addrResolved.size() > 0) {
+                addrDest = CAddress(addrResolved[GetRand(addrResolved.size())], NODE_NONE);
             }
         }
+    }
 
-        addrman.Attempt(addrConnect, fCountFailure);
+    if (!strDest.empty() && addrDest.IsValid()) {
+        // It is possible that we already have a connection to the IP/port pszDest resolved to.
+        // In that case, drop the connection that was just created, and return the existing CNode instead.
+        // Also store the name we used to connect in that CNode, so that future FindNode() calls to that
+        // name catch this early.
+        LOCK(cs_vNodes);
+        CNode* pnode = FindNode((CService)addrDest);
+        if (pnode)
+        {
+            pnode->MaybeSetAddrName(std::string(pszDest));
+            LogPrintf("Failed to open new connection, already connected\n");
+            return false;
+        }
+    }
 
-        // Add node
-        NodeId id = GetNewNodeId();
-        uint64_t nonce = GetDeterministicRandomizer(RANDOMIZER_ID_LOCALHOSTNONCE).Write(id).Finalize();
-        CNode* pnode = new CNode(id, nLocalServices, GetBestHeight(), hSocket, addrConnect, CalculateKeyedNetGroup(addrConnect), nonce, pszDest ? pszDest : "", false);
-        pnode->nServicesExpected = ServiceFlags(addrConnect.nServices & nRelevantServices);
-        pnode->AddRef();
+    // Configure proxy, if any. At this point, addrDest is resolved and valid unless we're connecting through
+    // a proxy which will be handling the resolve.
+    proxyType proxy;
+    if (addrDest.IsValid()) {
+        GetProxy(addrDest.GetNetwork(), proxy);
+    } else if (!GetNameProxy(proxy)) {
+        return false;
+    }
 
-        return pnode;
-    } else if (!proxyConnectionFailed) {
-        // If connecting to the node failed, and failure is not caused by a problem connecting to
-        // the proxy, mark this as an attempt.
-        addrman.Attempt(addrConnect, fCountFailure);
+    // Create socket and make sure that it is closed in the event of an error
+    SOCKET hSocket = INVALID_SOCKET;
+    CSocketCloser sockCloser(hSocket);
```

Comment:
> It seems strange to create the CSocketCloser and then give it a reference to the socket

It is strange, just like CloseSocket taking a reference is strange. But I'd think as long as one of these is taking a reference, the other probably should as well for consistency.